<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Wrapper around various existing tools and command-line interfaces,
  providing a standard interface, simple parallelization, and detailed logging.
  For microarray data, maps probe sets to standard gene IDs, building on
  GEOquery Davis and Meltzer (2007) &lt;doi:10.1093/bioinformatics/btm254&gt;,
  ArrayExpress Kauffmann et al. (2009) &lt;doi:10.1093/bioinformatics/btp354&gt;,
  Robust multi-array average RMA Irizarry et al. (2003) &lt;doi:10.1093/biostatistics/4.2.249&gt;,
  and BrainArray Dai et al. (2005) &lt;doi:10.1093/nar/gni179&gt;.
  For RNA-seq data, fetches metadata and raw reads from National Center for Biotechnology
  Information (NCBI) Sequence Read Archive (SRA), performs standard adapter and
  quality trimming using TrimGalore Krueger &lt;https://github.com/FelixKrueger/TrimGalore&gt;,
  performs quality control checks using FastQC Andrews &lt;https://github.com/s-andrews/FastQC&gt;,
  quantifies transcript abundances using salmon Patro et al. (2017) &lt;doi:10.1038/nmeth.4197&gt; and potentially
  refgenie Stolarczyk et al. (2020) &lt;doi:10.1093/gigascience/giz149&gt;,
  aggregates the results using MultiQC Ewels et al. (2016) &lt;doi:10.1093/bioinformatics/btw354&gt;,
  maps transcripts to genes using biomaRt Durinkck et al. (2009) &lt;doi:10.1038/nprot.2009.97&gt;,
  and summarizes transcript-level quantifications for gene-level analyses using
  tximport Soneson et al. (2015) &lt;doi:10.12688/f1000research.7563.2&gt;.">
<title>Simplified Fetching and Processing of Microarray and RNA-Seq Data • seeker</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Simplified Fetching and Processing of Microarray and RNA-Seq Data">
<meta property="og:description" content="Wrapper around various existing tools and command-line interfaces,
  providing a standard interface, simple parallelization, and detailed logging.
  For microarray data, maps probe sets to standard gene IDs, building on
  GEOquery Davis and Meltzer (2007) &lt;doi:10.1093/bioinformatics/btm254&gt;,
  ArrayExpress Kauffmann et al. (2009) &lt;doi:10.1093/bioinformatics/btp354&gt;,
  Robust multi-array average RMA Irizarry et al. (2003) &lt;doi:10.1093/biostatistics/4.2.249&gt;,
  and BrainArray Dai et al. (2005) &lt;doi:10.1093/nar/gni179&gt;.
  For RNA-seq data, fetches metadata and raw reads from National Center for Biotechnology
  Information (NCBI) Sequence Read Archive (SRA), performs standard adapter and
  quality trimming using TrimGalore Krueger &lt;https://github.com/FelixKrueger/TrimGalore&gt;,
  performs quality control checks using FastQC Andrews &lt;https://github.com/s-andrews/FastQC&gt;,
  quantifies transcript abundances using salmon Patro et al. (2017) &lt;doi:10.1038/nmeth.4197&gt; and potentially
  refgenie Stolarczyk et al. (2020) &lt;doi:10.1093/gigascience/giz149&gt;,
  aggregates the results using MultiQC Ewels et al. (2016) &lt;doi:10.1093/bioinformatics/btw354&gt;,
  maps transcripts to genes using biomaRt Durinkck et al. (2009) &lt;doi:10.1038/nprot.2009.97&gt;,
  and summarizes transcript-level quantifications for gene-level analyses using
  tximport Soneson et al. (2015) &lt;doi:10.12688/f1000research.7563.2&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">seeker</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.9</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/introduction.html">Introduction to seeker</a>
    <a class="dropdown-item" href="articles/reproducibility.html">Reproducibility with seeker</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.hugheylab.org/software/" target="_blank" aria-label="Hughey Lab">Hughey Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/hugheylab/seeker/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="seeker">seeker<a class="anchor" aria-label="anchor" href="#seeker"></a>
</h1></div>

<p><code>seeker</code> is an R package for fetching and processing sequencing data, especially RNA-seq data, as well as microarray data. Hopefully it helps you get what you’re after, before the day you die.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="docker-image">Docker image<a class="anchor" aria-label="anchor" href="#docker-image"></a>
</h3>
<p><code>seeker</code> and its dependencies are available in a Docker image called <a href="https://github.com/hugheylab/socker" class="external-link">socker</a>, based on <a href="https://github.com/rocker-org/rocker-versioned2" class="external-link">rocker/rstudio</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull ghcr.io/hugheylab/socker</span></code></pre></div>
<p>If you don’t want to use Docker, you can follow the instructions below.</p>
</div>
<div class="section level3">
<h3 id="r-package">R package<a class="anchor" aria-label="anchor" href="#r-package"></a>
</h3>
<div class="section level4">
<h4 id="option-1-cran">Option 1: CRAN<a class="anchor" aria-label="anchor" href="#option-1-cran"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">'BiocManager'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'BiocManager'</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'seeker'</span><span class="op">)</span> <span class="co"># handles Bioconductor dependencies</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="option-2-hughey-lab-drat-repository">Option 2: Hughey Lab Drat Repository<a class="anchor" aria-label="anchor" href="#option-2-hughey-lab-drat-repository"></a>
</h4>
<ol style="list-style-type: decimal">
<li>
<p>Install <a href="https://cran.r-project.org/package=BiocManager" class="external-link"><code>BiocManager</code></a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">'BiocManager'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'BiocManager'</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>If you use RStudio, go to Tools → Global Options… → Packages → Add… (under Secondary repositories), then enter:</p>
<ul>
<li>Name: hugheylab</li>
<li>Url: <a href="https://hugheylab.github.io/drat/" class="external-link uri">https://hugheylab.github.io/drat/</a>
</li>
</ul>
<p>You only have to do this once. Then you can install or update the package by entering:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'seeker'</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, you can install or update the package by entering:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'seeker'</span>, site_repository <span class="op">=</span> <span class="st">'https://hugheylab.github.io/drat/'</span><span class="op">)</span></span></code></pre></div>
</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="system-dependencies">System dependencies<a class="anchor" aria-label="anchor" href="#system-dependencies"></a>
</h3>
<p>If not using the docker image, use <a href="https://seeker.hugheylab.org/reference/installsysdeps"><code>seeker::installSysDeps()</code></a> to install and configure the various programs required to fetch and process sequencing data.</p>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>For an introduction to the package, read the <a href="https://seeker.hugheylab.org/articles/introduction.html">vignette</a>. For more details, check out the <a href="https://seeker.hugheylab.org/reference/index.html">reference documentation</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=seeker" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/hugheylab/seeker/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2" class="external-link">GPL-2</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing seeker</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jake Hughey <br><small class="roles"> Author, maintainer </small>  </li>
<li>Josh Schoenbachler <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/hugheylab/seeker/actions" class="external-link"><img src="https://github.com/hugheylab/seeker/workflows/check-deploy/badge.svg" alt="check-deploy"></a></li>
<li><a href="https://codecov.io/gh/hugheylab/seeker" class="external-link"><img src="https://codecov.io/gh/hugheylab/seeker/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://app.netlify.com/sites/hardcore-aryabhata-980673/deploys" class="external-link"><img src="https://api.netlify.com/api/v1/badges/a3d002cf-ca5c-427f-9963-061d282b1d1b/deploy-status" alt="Netlify Status"></a></li>
<li><a href="https://cran.r-project.org/package=seeker" class="external-link"><img src="https://www.r-pkg.org/badges/version/seeker" alt="CRAN Status"></a></li>
<li><a href="https://github.com/hugheylab/drat/tree/gh-pages/src/contrib" class="external-link"><img src="https://raw.githubusercontent.com/hugheylab/drat/gh-pages/badges/seeker_drat_badge.svg" alt="drat version"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jake Hughey, Josh Schoenbachler.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
