<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Simplified Fetching and Processing of Microarray and RNA-Seq Data • seeker</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Simplified Fetching and Processing of Microarray and RNA-Seq Data">
<meta name="description" content="Wrapper around various existing tools and command-line interfaces, providing a standard interface, simple parallelization, and detailed logging. For microarray data, maps probe sets to standard gene IDs, building on GEOquery Davis and Meltzer (2007) &lt;doi:10.1093/bioinformatics/btm254&gt;, ArrayExpress Kauffmann et al. (2009) &lt;doi:10.1093/bioinformatics/btp354&gt;, Robust multi-array average RMA Irizarry et al. (2003) &lt;doi:10.1093/biostatistics/4.2.249&gt;, and BrainArray Dai et al. (2005) &lt;doi:10.1093/nar/gni179&gt;. For RNA-seq data, fetches metadata and raw reads from National Center for Biotechnology Information (NCBI) Sequence Read Archive (SRA), performs standard adapter and quality trimming using TrimGalore Krueger &lt;https://github.com/FelixKrueger/TrimGalore&gt;, performs quality control checks using FastQC Andrews &lt;https://github.com/s-andrews/FastQC&gt;, quantifies transcript abundances using salmon Patro et al. (2017) &lt;doi:10.1038/nmeth.4197&gt; and potentially refgenie Stolarczyk et al. (2020) &lt;doi:10.1093/gigascience/giz149&gt;, aggregates the results using MultiQC Ewels et al. (2016) &lt;doi:10.1093/bioinformatics/btw354&gt;, maps transcripts to genes using biomaRt Durinkck et al. (2009) &lt;doi:10.1038/nprot.2009.97&gt;, and summarizes transcript-level quantifications for gene-level analyses using tximport Soneson et al. (2015) &lt;doi:10.12688/f1000research.7563.2&gt;.">
<meta property="og:description" content="Wrapper around various existing tools and command-line interfaces, providing a standard interface, simple parallelization, and detailed logging. For microarray data, maps probe sets to standard gene IDs, building on GEOquery Davis and Meltzer (2007) &lt;doi:10.1093/bioinformatics/btm254&gt;, ArrayExpress Kauffmann et al. (2009) &lt;doi:10.1093/bioinformatics/btp354&gt;, Robust multi-array average RMA Irizarry et al. (2003) &lt;doi:10.1093/biostatistics/4.2.249&gt;, and BrainArray Dai et al. (2005) &lt;doi:10.1093/nar/gni179&gt;. For RNA-seq data, fetches metadata and raw reads from National Center for Biotechnology Information (NCBI) Sequence Read Archive (SRA), performs standard adapter and quality trimming using TrimGalore Krueger &lt;https://github.com/FelixKrueger/TrimGalore&gt;, performs quality control checks using FastQC Andrews &lt;https://github.com/s-andrews/FastQC&gt;, quantifies transcript abundances using salmon Patro et al. (2017) &lt;doi:10.1038/nmeth.4197&gt; and potentially refgenie Stolarczyk et al. (2020) &lt;doi:10.1093/gigascience/giz149&gt;, aggregates the results using MultiQC Ewels et al. (2016) &lt;doi:10.1093/bioinformatics/btw354&gt;, maps transcripts to genes using biomaRt Durinkck et al. (2009) &lt;doi:10.1038/nprot.2009.97&gt;, and summarizes transcript-level quantifications for gene-level analyses using tximport Soneson et al. (2015) &lt;doi:10.12688/f1000research.7563.2&gt;.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">seeker</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/introduction.html">Introduction to seeker</a></li>
    <li><a class="dropdown-item" href="articles/reproducibility.html">Reproducibility with seeker</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://www.hugheylab.org/software/" target="_blank" aria-label="Hughey Lab">Hughey Lab</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/hugheylab/seeker/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="seeker">seeker<a class="anchor" aria-label="anchor" href="#seeker"></a>
</h1></div>
<p><a href="https://github.com/hugheylab/seeker/actions" class="external-link"><img src="https://github.com/hugheylab/seeker/workflows/check-deploy/badge.svg" alt="check-deploy"></a> <a href="https://codecov.io/gh/hugheylab/seeker" class="external-link"><img src="https://codecov.io/gh/hugheylab/seeker/branch/master/graph/badge.svg" alt="codecov"></a> <a href="https://app.netlify.com/sites/hardcore-aryabhata-980673/deploys" class="external-link"><img src="https://api.netlify.com/api/v1/badges/a3d002cf-ca5c-427f-9963-061d282b1d1b/deploy-status" alt="Netlify Status"></a> <a href="https://cran.r-project.org/package=seeker" class="external-link"><img src="https://www.r-pkg.org/badges/version/seeker" alt="CRAN Status"></a> <a href="https://github.com/hugheylab/drat/tree/gh-pages/src/contrib" class="external-link"><img src="https://raw.githubusercontent.com/hugheylab/drat/gh-pages/badges/seeker_drat_badge.svg" alt="drat version"></a></p>
<p><code>seeker</code> is an R package for fetching and processing sequencing data, especially RNA-seq data, as well as microarray data. Hopefully it helps you get what you’re after, before the day you die. For more details, see <a href="https://doi.org/10.7717/peerj.14372" class="external-link">Schoenbachler and Hughey (2022)</a>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To use <code>seeker</code>, you can install <code>seeker</code> and its dependencies, or use a pre-built Docker image in which <code>seeker</code> and its dependencies are already installed.</p>
<div class="section level3">
<h3 id="r-package-and-dependencies">R package and dependencies<a class="anchor" aria-label="anchor" href="#r-package-and-dependencies"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<p>Install the <code>BiocManager</code> R package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">'BiocManager'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'BiocManager'</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>Install the <code>seeker</code> R package, either from CRAN or from the Hughey Lab drat repository. We use <code><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">BiocManager::install()</a></code> in order to smoothly install <code>seeker</code>’s dependencies that are on Bioconductor.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'seeker'</span><span class="op">)</span> <span class="co"># CRAN</span></span>
<span><span class="co"># BiocManager::install('seeker', site_repository = 'https://hugheylab.github.io/drat/') # drat</span></span></code></pre></div>
</li>
<li>
<p>Install the system dependencies to fetch and process sequencing data. The simplest way to do this is to use the function <a href="https://seeker.hugheylab.org/reference/installsysdeps"><code>installSysDeps()</code></a>. For example,</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seeker</span><span class="fu">::</span><span class="fu"><a href="reference/installSysDeps.html">installSysDeps</a></span><span class="op">(</span><span class="st">'~'</span>, <span class="st">'~'</span>, <span class="st">'~'</span>, <span class="st">'~'</span><span class="op">)</span></span></code></pre></div>
<p>You can also use <code><a href="reference/installSysDeps.html">installSysDeps()</a></code> to fetch genomes from refgenie, such as those required to quantify transcript abundances using salmon.</p>
</li>
</ol>
</div>
<div class="section level3">
<h3 id="docker-image">Docker image<a class="anchor" aria-label="anchor" href="#docker-image"></a>
</h3>
<p>The Docker image is called <a href="https://github.com/hugheylab/socker" class="external-link">socker</a> and is based on <a href="https://github.com/rocker-org/rocker-versioned2" class="external-link">rocker/rstudio</a>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">docker</span> pull ghcr.io/hugheylab/socker</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>For an introduction to the package, read the <a href="https://seeker.hugheylab.org/articles/introduction.html">vignette</a>. For more details, check out the <a href="https://seeker.hugheylab.org/reference/index.html">reference documentation</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=seeker" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/hugheylab/seeker/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing seeker</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jake Hughey <br><small class="roles"> Author, maintainer </small>  </li>
<li>Josh Schoenbachler <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jake Hughey, Josh Schoenbachler.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
