<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Process genomic data • seeker</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Process genomic data">
<meta property="og:description" content="Simplified processing of sequencing and microarray data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">seeker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9048</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/introduction.html">Introduction to seeker</a>
    </li>
    <li>
      <a href="articles/reproducibility.html">Reproducibility with seeker</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hugheylab/seeker/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="seeker">seeker<a class="anchor" aria-label="anchor" href="#seeker"></a>
</h1></div>

<p><code>seeker</code> is an R package for fetching and processing sequencing data, especially RNA-seq data, as well as microarray data. Hopefully it helps you get what you’re after, before the day you die.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="docker-image">Docker image<a class="anchor" aria-label="anchor" href="#docker-image"></a>
</h3>
<p><code>seeker</code> and its dependencies are available in a Docker image called <a href="https://github.com/hugheylab/socker" class="external-link">socker</a>, based on <a href="https://github.com/rocker-org/rocker-versioned2" class="external-link">rocker/rstudio</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="ex">docker</span> pull ghcr.io/hugheylab/socker</span></code></pre></div>
<p>If you don’t want to use Docker, you can follow the instructions below.</p>
</div>
<div class="section level3">
<h3 id="r-package">R package<a class="anchor" aria-label="anchor" href="#r-package"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<p>Install <a href="https://cran.r-project.org/package=BiocManager" class="external-link"><code>BiocManager</code></a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">'BiocManager'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'BiocManager'</span><span class="op">)</span></code></pre></div>
</li>
<li>
<p>If you use RStudio, go to Tools → Global Options… → Packages → Add… (under Secondary repositories), then enter:</p>
<ul>
<li>Name: hugheylab</li>
<li>Url: <a href="https://hugheylab.github.io/drat/" class="external-link uri">https://hugheylab.github.io/drat/</a>
</li>
</ul>
<p>You only have to do this once. Then you can install or update the package by entering:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'seeker'</span><span class="op">)</span></code></pre></div>
<p>Alternatively, you can install or update the package by entering:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'seeker'</span>, site_repository <span class="op">=</span> <span class="st">'https://hugheylab.github.io/drat/'</span><span class="op">)</span></code></pre></div>
</li>
</ol>
</div>
<div class="section level3">
<h3 id="system-dependencies">System dependencies<a class="anchor" aria-label="anchor" href="#system-dependencies"></a>
</h3>
<p>The system dependencies are required for processing sequencing data. These instructions are for Unix-based systems, primarily Linux and macOS. If you’re using Windows, you’re doing it wrong.</p>
<ol style="list-style-type: decimal">
<li><p>Download and install <a href="https://www.ibm.com/aspera/connect/" class="external-link">Aspera Connect</a>. On Linux, you will likely have to download a tar.gz file (using <code>wget</code> or <code>curl</code>), untar it (using <code>tar -zxvf</code>), then run the resulting shell script. On macOS, you may have to install a browser extension first, then install Connect from a dmg file.</p></li>
<li>
<p>Install <a href="https://conda.io/en/latest/miniconda.html" class="external-link">Miniconda</a>. Use the default install location (~/miniconda3).</p>
<p>On Linux:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="ex">curl</span> -O https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="fu">sh</span> Miniconda3-latest-Linux-x86_64.sh</span></code></pre></div>
<p>On macOS:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">curl</span> -O https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="fu">sh</span> Miniconda3-latest-MacOSX-x86_64.sh</span></code></pre></div>
</li>
<li>
<p>Set up conda channels, including <a href="https://bioconda.github.io/user/install.html" class="external-link">bioconda</a>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">conda</span> config --add channels defaults</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="ex">conda</span> config --add channels bioconda</span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="ex">conda</span> config --add channels conda-forge</span></code></pre></div>
</li>
<li>
<p>Install the <a href="https://github.com/mamba-org/mamba" class="external-link">mamba package manager</a>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">conda</span> install mamba -c conda-forge</span></code></pre></div>
</li>
<li>
<p>Install the other command-line tools. The code below will install the packages into the base conda environment. If you use an environment other than base, please see the last step of this readme.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="ex">mamba</span> install refgenie trim-galore fastqc fastq-screen salmon multiqc</span></code></pre></div>
</li>
<li>
<p>Optionally, configure <a href="http://refgenie.databio.org/en/latest/install" class="external-link">refgenie</a>. For example:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="fu">mkdir</span> -p <span class="va">${HOME}</span>/genomes</span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="ex">refgenie</span> init -c <span class="va">${HOME}</span>/genomes/genome_config.yaml</span></code></pre></div>
<p>Then you can add the following line to ~/.bashrc, ~/.bash_profile, or ~/.zshrc, depending on your OS and shell.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1"></a><span class="bu">export</span> <span class="va">REFGENIE=</span><span class="st">"</span><span class="va">${HOME}</span><span class="st">/genomes/genome_config.yaml"</span></span></code></pre></div>
<p>Then <code>source</code> the file.</p>
</li>
<li>
<p>Optionally, use refgenie to fetch the salmon index files for the human and/or mouse transcriptomes.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1"></a><span class="ex">refgenie</span> pull hg38/salmon_sa_index</span>
<span id="cb12-2"><a href="#cb12-2"></a><span class="ex">refgenie</span> pull mm10/salmon_sa_index</span></code></pre></div>
</li>
<li>
<p>Optionally, fetch the genomes for fastq-screen. This takes a long time, so don’t bother unless you actually plan to run fastq-screen.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1"></a><span class="ex">fastq_screen</span> --get_genomes</span></code></pre></div>
</li>
<li>
<p>If you install Miniconda at a non-default location or install the system dependencies to a conda environment other than base, you will need to set a global option in each R session so <code>seeker</code> knows where to look. For example, if the dependencies are in an environment called “seeker” that <code>conda env list</code> says is located at ~/miniconda3/envs/seeker, then you should do the following:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>seeker.miniconda <span class="op">=</span> <span class="st">'~/miniconda3/envs/seeker'</span><span class="op">)</span></code></pre></div>
</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>For an introduction to the package, read the <a href="https://seeker.hugheylab.org/articles/introduction.html">vignette</a>. For more details, check out the <a href="https://seeker.hugheylab.org/reference/index.html">reference documentation</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/hugheylab/seeker/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2" class="external-link">GPL-2</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing seeker</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jake Hughey <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/hugheylab/seeker/actions" class="external-link"><img src="https://github.com/hugheylab/seeker/workflows/check-deploy/badge.svg" alt="check-deploy"></a></li>
<li><a href="https://codecov.io/gh/hugheylab/seeker" class="external-link"><img src="https://codecov.io/gh/hugheylab/seeker/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://app.netlify.com/sites/hardcore-aryabhata-980673/deploys" class="external-link"><img src="https://api.netlify.com/api/v1/badges/a3d002cf-ca5c-427f-9963-061d282b1d1b/deploy-status" alt="Netlify Status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Jake Hughey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.0.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
