<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function selectively performs various steps to process RNA-seq data.
See also the vignettes: browseVignettes('seeker')."><title>Process RNA-seq data end to end — seeker • seeker</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Process RNA-seq data end to end — seeker"><meta property="og:description" content="This function selectively performs various steps to process RNA-seq data.
See also the vignettes: browseVignettes('seeker')."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">seeker</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/introduction.html">Introduction to seeker</a>
    <a class="dropdown-item" href="../articles/reproducibility.html">Reproducibility with seeker</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://www.hugheylab.org/software/" target="_blank" aria-label="Hughey Lab">Hughey Lab</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/hugheylab/seeker/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Process RNA-seq data end to end</h1>
      <small class="dont-index">Source: <a href="https://github.com/hugheylab/seeker/blob/HEAD/R/seeker.R" class="external-link"><code>R/seeker.R</code></a></small>
      <div class="d-none name"><code>seeker.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function selectively performs various steps to process RNA-seq data.
See also the vignettes: <code>browseVignettes('seeker')</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">seeker</span><span class="op">(</span><span class="va">params</span>, parentDir <span class="op">=</span> <span class="st">"."</span>, dryRun <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>params</dt>
<dd><p>Named list of parameters with components:</p><ul><li><p><code>study</code>: String used to name the output directory within <code>parentDir</code>.</p></li>
<li><p><code>metadata</code>: Named list with components:</p><ul><li><p><code>run</code>: Logical indicating whether to fetch metadata. See
<code><a href="fetchMetadata.html">fetchMetadata()</a></code>. If <code>TRUE</code>, saves a file
<code>parentDir</code>/<code>study</code>/metadata.csv. If <code>FALSE</code>, expects that file to
already exist. The unmodified fetched or found metadata is saved to a
file <code>parentDir</code>/<code>study</code>/metadata_original.csv. Following components
are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>bioproject</code>: String indicating the study's bioproject accession.</p></li>
<li><p><code>include</code>: Optional named list for specifying which rows of metadata to
include for further processing, with components:</p><ul><li><p><code>colname</code>: String indicating column in metadata</p></li>
<li><p><code>values</code>: Vector indicating values within <code>colname</code></p></li>
</ul></li>
<li><p><code>exclude</code>: Optional named list for specifying which rows of metadata to
exclude from further processing (superseding <code>include</code>), with components:</p><ul><li><p><code>colname</code>: String indicating column in metadata</p></li>
<li><p><code>values</code>: Vector indicating values within <code>colname</code></p></li>
</ul></li>
</ul></li>
<li><p><code>fetch</code>: Named list with components:</p><ul><li><p><code>run</code>: Logical indicating whether to fetch files from SRA. See <code><a href="fetch.html">fetch()</a></code>.
If <code>TRUE</code>, saves files to <code>parentDir</code>/<code>study</code>/fetch_output. Whether
<code>TRUE</code> or <code>FALSE</code>, expects metadata to have a column "run_accession", and
updates metadata with column "fastq_fetched" containing paths to files in
<code>parentDir</code>/<code>study</code>/fetch_output. Following components are only checked
if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>keep</code>: Logical indicating whether to keep fastq.gz files when all
processing steps have completed. <code>NULL</code> indicates <code>TRUE</code>.</p></li>
<li><p><code>overwrite</code>: Logical indicating whether to overwrite files that already
exist. <code>NULL</code> indicates to use the default in <code><a href="fetch.html">fetch()</a></code>.</p></li>
<li><p><code>keepSra</code>: Logical indicating whether to keep the ".sra" files. <code>NULL</code>
indicates to use the default in <code><a href="fetch.html">fetch()</a></code>.</p></li>
<li><p><code>prefetchCmd</code>: String indicating command for prefetch, which downloads
".sra" files. <code>NULL</code> indicates to use the default in <code><a href="fetch.html">fetch()</a></code>.</p></li>
<li><p><code>prefetchArgs</code>: Character vector indicating arguments to pass to
prefetch. <code>NULL</code> indicates to use the default in <code><a href="fetch.html">fetch()</a></code>.</p></li>
<li><p><code>fasterqdumpCmd</code>: String indicating command for fasterq-dump, which
uses ".sra" files to create ".fastq" files. <code>NULL</code> indicates to use the
default in <code><a href="fetch.html">fetch()</a></code>.</p></li>
<li><p><code>prefetchArgs</code>: Character vector indicating arguments to pass to
fasterq-dump. <code>NULL</code> indicates to use the default in <code><a href="fetch.html">fetch()</a></code>.</p></li>
<li><p><code>pigzCmd</code>: String indicating command for pigz, which converts ".fastq"
files to ".fastq.gz" files. <code>NULL</code> indicates to use the default in
<code><a href="fetch.html">fetch()</a></code>.</p></li>
<li><p><code>pigzArgs</code>: Character vector indicating arguments to pass to pigz. <code>NULL</code>
indicates to use the default in <code><a href="fetch.html">fetch()</a></code>.</p></li>
</ul></li>
<li><p><code>trimgalore</code>: Named list with components:</p><ul><li><p><code>run</code>: Logical indicating whether to perform quality/adapter trimming of
reads. See <code><a href="trimgalore.html">trimgalore()</a></code>. If <code>TRUE</code>, expects metadata to have a column
"fastq_fetched" containing paths to fastq files in
<code>parentDir</code>/<code>study</code>/fetch_output, saves trimmed files to
<code>parentDir</code>/<code>study</code>/trimgalore_output, and updates metadata with column
"fastq_trimmed". If <code>FALSE</code>, expects and does nothing. Following
components are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>keep</code>: Logical indicating whether to keep trimmed fastq files when all
processing steps have completed. <code>NULL</code> indicates <code>TRUE</code>.</p></li>
<li><p><code>cmd</code>: Name or path of the command-line interface. <code>NULL</code> indicates to
use the default in <code><a href="trimgalore.html">trimgalore()</a></code>.</p></li>
<li><p><code>args</code>: Additional arguments to pass to the command-line interface.
<code>NULL</code> indicates to use the default in <code><a href="trimgalore.html">trimgalore()</a></code>.</p></li>
<li><p><code>pigzCmd</code>: String indicating command for pigz, which converts ".fastq"
files to ".fastq.gz" files. <code>NULL</code> indicates to use the default in
<code><a href="trimgalore.html">trimgalore()</a></code>.</p></li>
</ul></li>
<li><p><code>fastqc</code>: Named list with components:</p><ul><li><p><code>run</code>: Logical indicating whether to perform QC on reads. See <code><a href="fastqc.html">fastqc()</a></code>.
If <code>TRUE</code> and <code>trimgalore$run</code> is <code>TRUE</code>, expects metadata to have a
column "fastq_trimmed" containing paths to fastq files in
<code>parentDir</code>/<code>study</code>/trimgalore_output. If <code>TRUE</code> and <code>trimgalore$run</code> is
<code>FALSE</code>, expects metadata to have a column "fastq_fetched" containing
paths to fastq files in <code>parentDir</code>/<code>study</code>/fetch_output. If <code>TRUE</code>,
saves results to <code>parentDir</code>/<code>study</code>/fastqc_output. If <code>FALSE</code>, expects
and does nothing. Following components are only checked if <code>run</code> is
<code>TRUE</code>.</p></li>
<li><p><code>keep</code>: Logical indicating whether to keep fastqc files when all
processing steps have completed. <code>NULL</code> indicates <code>TRUE</code>.</p></li>
<li><p><code>cmd</code>: Name or path of the command-line interface. <code>NULL</code> indicates to
use the default in <code><a href="fastqc.html">fastqc()</a></code>.</p></li>
<li><p><code>args</code>: Additional arguments to pass to the command-line interface.
<code>NULL</code> indicates to use the default in <code><a href="fastqc.html">fastqc()</a></code>.</p></li>
</ul></li>
<li><p><code>salmon</code>: Named list with components:</p><ul><li><p><code>run</code>: Logical indicating whether to quantify transcript abundances. See
<code><a href="salmon.html">salmon()</a></code>. If <code>TRUE</code> and <code>trimgalore$run</code> is <code>TRUE</code>, expects metadata to
have a column "fastq_trimmed" containing paths to fastq files in
<code>parentDir</code>/<code>study</code>/trimgalore_output. If <code>TRUE</code> and <code>trimgalore$run</code> is
<code>FALSE</code>, expects metadata to have a column "fastq_fetched" containing
paths to fastq files in <code>parentDir</code>/<code>study</code>/fetch_output. If <code>TRUE</code>,
saves results to <code>parentDir</code>/<code>study</code>/salmon_output and
<code>parentDir</code>/<code>study</code>/salmon_meta_info.csv. If <code>FALSE</code>, expects and does
nothing. Following components are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>indexDir</code>: Directory that contains salmon index.</p></li>
<li><p><code>sampleColname</code>: String indicating column in metadata containing sample
ids. <code>NULL</code> indicates "sample_accession", which should work for data
from SRA and ENA.</p></li>
<li><p><code>keep</code>: Logical indicating whether to keep quantification results when
all processing steps have completed. <code>NULL</code> indicates <code>TRUE</code>.</p></li>
<li><p><code>cmd</code>: Name or path of the command-line interface. <code>NULL</code> indicates to
use the default in <code><a href="salmon.html">salmon()</a></code>.</p></li>
<li><p><code>args</code>: Additional arguments to pass to the command-line interface.
<code>NULL</code> indicates to use the default in <code><a href="salmon.html">salmon()</a></code>.</p></li>
</ul></li>
<li><p><code>multiqc</code>: Named list with components:</p><ul><li><p><code>run</code>: Logical indicating whether to aggregrate results of various
processing steps. See <code><a href="multiqc.html">multiqc()</a></code>. If <code>TRUE</code>, saves results to
<code>parentDir</code>/<code>study</code>/multiqc_output. If <code>FALSE</code>, expects and does nothing.
Following components are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>cmd</code>: Name or path of the command-line interface. <code>NULL</code> indicates to
use the default in <code><a href="multiqc.html">multiqc()</a></code>.</p></li>
<li><p><code>args</code>: Additional arguments to pass to the command-line interface.
<code>NULL</code> indicates to use the default in <code><a href="multiqc.html">multiqc()</a></code>.</p></li>
</ul></li>
<li><p><code>tximport</code>: Named list with components:</p><ul><li><p><code>run</code>: Logical indicating whether to summarize transcript- or gene-level
estimates for downstream analysis. See <code><a href="tximport.html">tximport()</a></code>. If <code>TRUE</code>, expects
metadata to have a column <code>sampleColname</code> of sample ids, and expects a
directory <code>parentDir</code>/<code>study</code>/salmon_output containing directories of
quantification results, and saves results to
<code>parentDir</code>/<code>study</code>/tximport_output.qs. If <code>FALSE</code>, expects and does
nothing. Following components are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>tx2gene</code>: Optional named list with components:</p><ul><li><p><code>organism</code>: String indicating organism and thereby ensembl gene dataset.
See <code><a href="getTx2gene.html">getTx2gene()</a></code>.</p></li>
<li><p><code>version</code>: Optional number indicating ensembl version. <code>NULL</code> indicates
the latest version. See <code><a href="getTx2gene.html">getTx2gene()</a></code>.</p></li>
<li><p><code>filename</code>: Optional string indicating name of pre-existing text file
in <code>parentDir</code>/<code>params$study</code> containing mapping between transcripts
(first column) and genes (second column), with column names in the
first row. If <code>filename</code> is specified, <code>organism</code> and <code>version</code> must not
be specified.</p></li>
</ul><p>If not <code>NULL</code>, saves a file <code>parentDir</code>/<code>study</code>/tx2gene.csv.gz.</p></li>
<li><p><code>countsFromAbundance</code>: String indicating whether or how to estimate
counts using estimated abundances. See <code><a href="https://rdrr.io/pkg/tximport/man/tximport.html" class="external-link">tximport::tximport()</a></code>.</p></li>
<li><p><code>ignoreTxVersion</code>: Logical indicating whether to the version suffix on
transcript ids. <code>NULL</code> indicates to use <code>TRUE</code>. See
<code><a href="https://rdrr.io/pkg/tximport/man/tximport.html" class="external-link">tximport::tximport()</a></code>.</p></li>
</ul></li>
</ul><p><code>params</code> can be derived from a yaml file, see
<code><a href="../articles/introduction.html">vignette("introduction", package = "seeker")</a></code>. The yaml representation
of <code>params</code> will be saved to <code>parentDir</code>/<code>params$study</code>/params.yml.</p></dd>


<dt>parentDir</dt>
<dd><p>Directory in which to store the output, which will be a
directory named according to <code>params$study</code>.</p></dd>


<dt>dryRun</dt>
<dd><p>Logical indicating whether to check the validity of inputs
without actually fetching or processing any data.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Path to the output directory <code>parentDir</code>/<code>params$study</code>, invisibly.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="fetchMetadata.html">fetchMetadata()</a></code>, <code><a href="fetch.html">fetch()</a></code>, <code><a href="trimgalore.html">trimgalore()</a></code>, <code><a href="fastqc.html">fastqc()</a></code>,
<code><a href="salmon.html">salmon()</a></code>, <code><a href="multiqc.html">multiqc()</a></code>, <code><a href="tximport.html">tximport()</a></code>, <code><a href="installSysDeps.html">installSysDeps()</a></code>, <code><a href="seekerArray.html">seekerArray()</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu">doParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html" class="external-link">registerDoParallel</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">params</span> <span class="op">=</span> <span class="fu">yaml</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/yaml/man/read_yaml.html" class="external-link">read_yaml</a></span><span class="op">(</span><span class="st">'my_params.yaml'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">seeker</span><span class="op">(</span><span class="va">params</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Jake Hughey, Josh Schoenbachler.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

