<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Process RNA-seq data end to end — seeker • seeker</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Process RNA-seq data end to end — seeker" />
<meta property="og:description" content="This function selectively performs various steps to process RNA-seq data." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seeker</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9033</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">Introduction to seeker</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hugheylab/seeker/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Process RNA-seq data end to end</h1>
    <small class="dont-index">Source: <a href='https://github.com/hugheylab/seeker/blob/master/R/seeker.R'><code>R/seeker.R</code></a></small>
    <div class="hidden name"><code>seeker.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function selectively performs various steps to process RNA-seq data.</p>
    </div>

    <pre class="usage"><span class='fu'>seeker</span><span class='op'>(</span><span class='va'>params</span>, parentDir <span class='op'>=</span> <span class='st'>"."</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>params</th>
      <td><p>Named list of parameters with components:</p><ul>
<li><p><code>study</code>: String used to name the output directory within <code>parentDir</code>.</p></li>
<li><p><code>metadata</code>: Named list with components:</p><ul>
<li><p><code>run</code>: Logical indicating whether to fetch metadata from ENA. See
<code><a href='fetchMetadata.html'>fetchMetadata()</a></code>. If <code>TRUE</code>, saves a file
<code>parentDir</code>/<code>study</code>/data/metadata.csv. If <code>FALSE</code>, expects that file to
already exist. Following components are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>bioproject</code>: String indicating the study's bioproject accession.</p></li>
<li><p><code>include</code>: Optional named list for specifying which rows of metadata to
include for further processing, with components:</p><ul>
<li><p><code>colname</code>: String indicating column in metadata</p></li>
<li><p><code>values</code>: Vector indicating values within <code>colname</code></p></li>
</ul></li>
<li><p><code>exclude</code>: Optional named list for specifying which rows of metadata to
exclude from further processing (superseding <code>include</code>), with components:</p><ul>
<li><p><code>colname</code>: String indicating column in metadata</p></li>
<li><p><code>values</code>: Vector indicating values within <code>colname</code></p></li>
</ul></li>
</ul></li>
<li><p><code>fetch</code>: Named list with components:</p><ul>
<li><p><code>run</code>: Logical indicating whether to fetch fastq(.gz) files using ascp.
See <code><a href='fetch.html'>fetch()</a></code>. If <code>TRUE</code>, expects metadata to have a column
'fastq_aspera' containing remote paths, and saves files to
<code>parentDir</code>/<code>study</code>/fetch_output. If <code>FALSE</code>, expects metadata to have a
column 'fastq_aspera' containing names (or complete paths, local or
remote) of fastq files. Whether <code>TRUE</code> or <code>FALSE</code>, updates metadata with
column 'fastq_fetched' containing paths to files that should be in
<code>parentDir</code>/<code>study</code>/fetch_output. Following components are only checked
if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>keep</code>: Logical indicating whether to keep fetched fastq files when all
processing steps have completed. <code>NULL</code> indicates <code>TRUE</code>.</p></li>
<li><p><code>overwrite</code>: Logical indicating whether to overwrite files that already
exist. <code>NULL</code> indicates to use the default in <code><a href='fetch.html'>fetch()</a></code>.</p></li>
<li><p><code>ascpCmd</code>: String indicating path to ascp. <code>NULL</code> indicates to use the
default in <code><a href='fetch.html'>fetch()</a></code>.</p></li>
<li><p><code>ascpArgs</code>: Character vector of arguments to pass to ascp. <code>NULL</code>
indicates to use the default in <code><a href='fetch.html'>fetch()</a></code>.</p></li>
<li><p><code>ascpPrefix</code>: String indicating prefix for fetching files. <code>NULL</code>
indicates to use the default in <code><a href='fetch.html'>fetch()</a></code>.</p></li>
</ul></li>
<li><p><code>trimgalore</code>: Named list with components:</p><ul>
<li><p><code>run</code>: Logical indicating whether to perform quality/adapter trimming of
reads. See <code><a href='trimgalore.html'>trimgalore()</a></code>. If <code>TRUE</code>, expects metadata to have a column
'fastq_fetched' containing paths to fastq files in
<code>parentDir</code>/<code>study</code>/fetch_output, saves trimmed files to
<code>parentDir</code>/<code>study</code>/trimgalore_output, and updates metadata with column
'fastq_trimmed'. If <code>FALSE</code>, expects and does nothing. Following
components are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>keep</code>: Logical indicating whether to keep trimmed fastq files when all
processing steps have completed. <code>NULL</code> indicates <code>TRUE</code>.</p></li>
<li><p><code>cmd</code>: Name or path of the command-line interface. <code>NULL</code> indicates to
use the default in <code><a href='trimgalore.html'>trimgalore()</a></code>.</p></li>
<li><p><code>args</code>: Additional arguments to pass to the command-line interface.
<code>NULL</code> indicates to use the default in <code><a href='trimgalore.html'>trimgalore()</a></code>.</p></li>
</ul></li>
<li><p><code>fastqc</code>: Named list with components:</p><ul>
<li><p><code>run</code>: Logical indicating whether to perform QC on reads. See <code><a href='fastqc.html'>fastqc()</a></code>.
If <code>TRUE</code> and <code>trimgalore$run</code> is <code>TRUE</code>, expects metadata to have a
column 'fastq_trimmed' containing paths to fastq files in
<code>parentDir</code>/<code>study</code>/trimgalore_output. If <code>TRUE</code> and <code>trimgalore$run</code> is
<code>FALSE</code>, expects metadata to have a column 'fastq_fetched' containing
paths to fastq files in <code>parentDir</code>/<code>study</code>/fetch_output. If <code>TRUE</code>,
saves results to <code>parentDir</code>/<code>study</code>/fastqc_output. If <code>FALSE</code>, expects
and does nothing. Following components are only checked if <code>run</code> is
<code>TRUE</code>.</p></li>
<li><p><code>keep</code>: Logical indicating whether to keep fastqc files when all
processing steps have completed. <code>NULL</code> indicates <code>TRUE</code>.</p></li>
<li><p><code>cmd</code>: Name or path of the command-line interface. <code>NULL</code> indicates to
use the default in <code><a href='fastqc.html'>fastqc()</a></code>.</p></li>
<li><p><code>args</code>: Additional arguments to pass to the command-line interface.
<code>NULL</code> indicates to use the default in <code><a href='fastqc.html'>fastqc()</a></code>.</p></li>
</ul></li>
<li><p><code>salmon</code>: Named list with components:</p><ul>
<li><p><code>run</code>: Logical indicating whether to quantify transcript abundances. See
<code><a href='salmon.html'>salmon()</a></code>. If <code>TRUE</code> and <code>trimgalore$run</code> is <code>TRUE</code>, expects metadata to
have a column 'fastq_trimmed' containing paths to fastq files in
<code>parentDir</code>/<code>study</code>/trimgalore_output. If <code>TRUE</code> and <code>trimgalore$run</code> is
<code>FALSE</code>, expects metadata to have a column 'fastq_fetched' containing
paths to fastq files in <code>parentDir</code>/<code>study</code>/fetch_output. If <code>TRUE</code>, also
expects metadata to have a column 'sample_accession' containing sample
ids, and saves results to <code>parentDir</code>/<code>study</code>/salmon_output and
<code>parentDir</code>/<code>study</code>/data/salmon_meta_info.csv. If <code>FALSE</code>, expects and
does nothing. Following components are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>indexDir</code>: Directory that contains salmon index.</p></li>
<li><p><code>keep</code>: Logical indicating whether to keep quantification results when
all processing steps have completed. <code>NULL</code> indicates <code>TRUE</code>.</p></li>
<li><p><code>cmd</code>: Name or path of the command-line interface. <code>NULL</code> indicates to
use the default in <code><a href='salmon.html'>salmon()</a></code>.</p></li>
<li><p><code>args</code>: Additional arguments to pass to the command-line interface.
<code>NULL</code> indicates to use the default in <code><a href='salmon.html'>salmon()</a></code>.</p></li>
</ul></li>
<li><p><code>multiqc</code>: Named list with components:</p><ul>
<li><p><code>run</code>: Logical indicating whether to aggregrate results of various
processing steps. See <code><a href='multiqc.html'>multiqc()</a></code>. If <code>TRUE</code>, saves results to
<code>parentDir</code>/<code>study</code>/multiqc_output. If <code>FALSE</code>, expects and does nothing.
Following components are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>cmd</code>: Name or path of the command-line interface. <code>NULL</code> indicates to
use the default in <code><a href='multiqc.html'>multiqc()</a></code>.</p></li>
<li><p><code>args</code>: Additional arguments to pass to the command-line interface.
<code>NULL</code> indicates to use the default in <code><a href='multiqc.html'>multiqc()</a></code>.</p></li>
</ul></li>
<li><p><code>tximport</code>: Named list with components:</p><ul>
<li><p><code>run</code>: Logical indicating whether to summarize transcript- or gene-level
estimates for downstream analysis. See <code><a href='tximport.html'>tximport()</a></code>. If <code>TRUE</code>, expects
metadata to have a column 'sample_accession' of sample ids, and expects a
directory <code>parentDir</code>/<code>study</code>/salmon_output containing directories of
quantification results, and saves results to
<code>parentDir</code>/<code>study</code>/data/tximport_output.qs. If <code>FALSE</code>, expects and does
nothing. Following components are only checked if <code>run</code> is <code>TRUE</code>.</p></li>
<li><p><code>tx2gene</code>: Optional named list with components:</p><ul>
<li><p><code>species</code>: String indicating species and thereby ensembl gene dataset.
See <code><a href='getTx2gene.html'>getTx2gene()</a></code>.</p></li>
<li><p><code>version</code>: Optional number indicating ensembl version. <code>NULL</code> indicates
the latest version. See <code><a href='getTx2gene.html'>getTx2gene()</a></code>.</p></li>
</ul>

<p>If not <code>NULL</code>, saves a file <code>parentDir</code>/<code>study</code>/data/tx2gene.csv.gz.</p></li>
<li><p><code>countsFromAbundance</code>: String indicating whether or how to estimate
counts using estimated abundances. See <code><a href='https://rdrr.io/pkg/tximport/man/tximport.html'>tximport::tximport()</a></code>.</p></li>
<li><p><code>ignoreTxVersion</code>: Logical indicating whether to the version suffix on
transcript ids. <code>NULL</code> indicates to use <code>TRUE</code>. See
<code><a href='https://rdrr.io/pkg/tximport/man/tximport.html'>tximport::tximport()</a></code>.</p></li>
</ul></li>
</ul>

<p><code>params</code> can be derived from a yaml file, see
<code><a href='../articles/introduction.html'>vignette('introduction', package = 'seeker')</a></code>. The yaml representation
of <code>params</code> will be saved to <code>parentDir</code>/<code>params$study</code>/data/params.yml.</p></td>
    </tr>
    <tr>
      <th>parentDir</th>
      <td><p>Directory in which to store the output, which will be a
directory named according to <code>params$study</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>NULL</code>, invisibly.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='fetchMetadata.html'>fetchMetadata()</a></code>, <code><a href='fetch.html'>fetch()</a></code>, <code><a href='trimgalore.html'>trimgalore()</a></code>, <code><a href='fastqc.html'>fastqc()</a></code>,
<code><a href='salmon.html'>salmon()</a></code>, <code><a href='multiqc.html'>multiqc()</a></code>, <code><a href='tximport.html'>tximport()</a></code></p></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jake Hughey.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


